internal/controller/gatewayclass.go

```diff
@@ -0,0 +1,128 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/projectcontour/contour/internal/status"
+	"github.com/projectcontour/contour/internal/validation"
+	"github.com/projectcontour/contour/pkg/config"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gatewayapi_v1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
+)
+
+type gatewayClassReconciler struct {
+	client       client.Client
+	eventHandler cache.ResourceEventHandler
+	log          logrus.FieldLogger
+}
+
+// NewGatewayClassController creates the gatewayclass controller from mgr. The controller will be
+// pre-configured to watch for GatewayClass objects.
+func NewGatewayClassController(mgr manager.Manager, eventHandler cache.ResourceEventHandler, log logrus.FieldLogger) (controller.Controller, error) {
+	r := &gatewayClassReconciler{
+		client:       mgr.GetClient(),
+		eventHandler: eventHandler,
+		log:          log,
+	}
+
+	c, err := controller.New("gatewayclass-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return nil, err
+	}
+
+	// Only enqueue GatewayClass objects that specify contour as the controller.
+	if err := c.Watch(&source.Kind{Type: &gatewayapi_v1alpha1.GatewayClass{}}, r.enqueueRequestForGatewayClass()); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// enqueueRequestForGatewayClass returns an event handler that maps events to
+// GatewayClass objects owned by the controller.
+func (r *gatewayClassReconciler) enqueueRequestForGatewayClass() handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(a client.Object) []reconcile.Request {
+		gc, ok := a.(*gatewayapi_v1alpha1.GatewayClass)
+		if !ok {
+			r.log.WithField("name", a.GetName()).Info("invalid object, bypassing reconciliation.")
+			return []reconcile.Request{}
+		}
+		if gc.Spec.Controller == config.ContourGatewayClass {
+			r.log.WithField("name", gc.Name).Info("queueing gatewayclass")
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Name: gc.Name,
+					},
+				},
+			}
+		}
+		r.log.WithField("name", gc.Name).Info("gatewayclass not owned by contour, bypassing reconciliation")
```

Comment:
Suggest maybe adding the `config.ContourGatewayClass` string to the error to tell folks what it required to match?